### PR TITLE
perf(loudnorm): skip video decode in measurement and skip no-op normalization

### PR DESCRIPTION
**Pull request by christosporios**
## Problem

The two-pass loudness normalization added in d7185b9 runs its measurement pass over the full file without `-vn`, so ffmpeg decodes the entire video stream for a measurement that only reads audio. On staging this took **34 minutes at 150% CPU** for a 6h41m / 2.13 GB meeting — and logged nothing until done, making the pipeline look hung.

That same staging run then re-encoded the audio and re-muxed the whole file to correct a measured loudness of −14.46 LUFS against a −14 target — an inaudible 0.5 dB adjustment (though in that case the true peak did exceed the ceiling, so pass 2 was legitimate).

## Changes

- **Pass 1 adds `-vn`** so only the audio stream is decoded: the measurement drops from ~35 minutes to a couple of minutes on long meetings. A log line now marks the start of the pass.
- **Pass 2 is skipped** when the measured integrated loudness is within 1 LU of the target *and* the true peak is at or under the −1.5 dBTP ceiling (`needsLoudnormCorrection`, unit-tested). Unparseable measurements (e.g. `-inf` on silent audio) conservatively fall through to normalization, preserving the old behavior.
- Because normalized output measures at the target, reprocessing a cached video now skips the expensive pass entirely instead of re-encoding it every run.

## Testing

- `npm run typecheck` and `npm test` pass (399 tests, 6 new for `needsLoudnormCorrection`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Low Risk**
> Scoped ffmpeg pipeline optimization with conservative fallback on bad measurements; behavior change is skipping inaudible re-encodes and faster measurement only.
> 
> **Overview**
> Speeds up **two-pass loudness normalization** in the download pipeline and avoids pointless re-encodes on cached or already-normalized videos.
> 
> **Pass 1** now passes **`-vn`** before **`-i`** so ffmpeg measures loudness from audio only (no full video decode on long meetings), and logs when measurement starts. **Pass 2** is skipped when **`needsLoudnormCorrection`** sees integrated loudness within **1 LU** of the −14 LUFS target; **true peak is intentionally ignored** so AAC re-encode TP overshoot does not trigger endless re-runs. Non-finite measurements (e.g. `-inf`) still normalize. Six unit tests cover the new gate.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0537e7f8ec8b2f5d74bfec4a0c6c3756f5886a66. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Greptile Summary</h3>

Speeds up two-pass loudness normalization by measuring audio-only (pass 1 `-vn` now placed as an input-side option) and skipping pass 2 when integrated loudness is already within 1 LU of the −14 LUFS target, addressing a 34-minute measurement regression on long recordings.

- `needsLoudnormCorrection` deliberately omits a true-peak check: a production observation showed the AAC encoder overshoots the TP limiter on each normalize cycle, so a TP-only trigger would re-encode indefinitely without converging; this trade-off is well-documented in both code comments and the new test.
- Six targeted unit tests cover the skip/trigger boundary, exact-target, TP-overshoot-with-on-target-loudness, and non-finite (`-inf`) fall-through cases.
- Cached already-normalized files now exit after the quick measurement pass with no re-encode.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the changes are additive, well-tested, and the deliberate omission of a true-peak trigger is thoroughly explained and unit-tested to prevent an infinite re-encode loop.

Both changed paths (the measurement skip and the `-vn` input placement) work correctly. The `needsLoudnormCorrection` guard handles all edge cases — `-inf` on silent audio falls through to normalization, NaN from unparseable strings does the same, and the on-target fast-exit is covered by two explicit tests. The TP-only skipping design is intentionally conservative given the observed encoder overshoot, and is clearly documented.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/tasks/downloadYTV.ts | Adds `needsLoudnormCorrection` skip guard, moves `-vn` to input-side for pass 1, and adds an early-measurement log line. Logic is sound and well-commented. |
| src/tasks/downloadYTV.test.ts | Six new tests cover tolerance boundary, exact-target hit, TP-overshoot-with-on-target-loudness, too-quiet, too-loud, and non-finite measurement. All cases are distinct and the TP-overshoot test correctly documents the deliberate production trade-off. |

</details>

<sub>Reviews (3): Last reviewed commit: ["fix(loudnorm): only loudness deviation t..."](https://github.com/schemalabz/opencouncil-tasks/commit/0537e7f8ec8b2f5d74bfec4a0c6c3756f5886a66) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36902181)</sub>

### CODE REVIEW

File: src/tasks/downloadYTV.ts [hunk 1]
```diff
@@ -100,8 +100,22 @@ type LoudnormStats = {
...
+const LOUDNORM_TOLERANCE_LU = 1;
 const AUDIO_BITRATE = '128k';
 
+export function needsLoudnormCorrection(stats: LoudnormStats): boolean {
+    const deltaI = Math.abs(LOUDNORM_TARGET.I - parseFloat(stats.input_i));
+    const truePeak = parseFloat(stats.input_tp);
+    if (Number.isNaN(deltaI) || Number.isNaN(truePeak)) {
+        return true;
+    }
+    return deltaI > LOUDNORM_TOLERANCE_LU || truePeak > LOUDNORM_TARGET.TP;
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/tasks/downloadYTV.ts]: Fixed in 26b220b: the guard now uses `Number.isFinite` so any non-finite measurement falls through to normalization. (Note `parseFloat('-inf')` is actually `NaN`, not `-Infinity` — ffmpeg's `-inf` was already caught — but the finite check is strictly more robust and now covered by a test.)

File: src/tasks/downloadYTV.ts [hunk 2]
```diff
@@ -149,10 +163,14 @@ async function normalizeVideoAudio(filePath: string): Promise<void> {
...
+    // stream, which would otherwise dominate the runtime of a measurement
+    // that only reads audio (~35 minutes of video decode for a 6.5h meeting).
+    console.log(`[loudnorm] ${filename}: measuring loudness...`);
     const pass1Start = Date.now();
     const { stderr } = await runFfmpeg([
         '-i', filePath,
+        '-vn',
         '-af', `${loudnormFilter}:print_format=json`,
         '-f', 'null', '-',
     ]);
```
[the remark this answers is not in the document]
Comment (christosporios) [reply src/tasks/downloadYTV.ts:184]: Done in 26b220b — moved `-vn` before `-i` as an input option. Verified against the bundled ffmpeg-static binary that the loudnorm JSON output is unchanged.

**Comment by christosporios** [comment]
Field report from staging led to one more fix: the skip predicate originally re-normalized when true peak exceeded the −1.5 dBTP ceiling. In practice the AAC re-encode in pass 2 *overshoots* the limiter — one normalize cycle measured TP going +0.69 → **+1.98 dBTP** while loudness stayed on target (−14.46 → −14.64 LUFS). So a TP-only trigger can never converge: every rerun of a cached video re-normalized for ~90 minutes and added a lossy generation. The predicate now triggers on integrated-loudness deviation only (TP correction through a lossy re-encode is demonstrably ineffective). Tests updated with the real measurements.

File: src/tasks/downloadYTV.ts [hunk 3]
```diff
@@ -162,6 +188,11 @@ async function normalizeVideoAudio(filePath: string): Promise<void> {
     const deltaI = (I - parseFloat(stats.input_i)).toFixed(1);
     console.log(`[loudnorm] ${filename}: I=${stats.input_i} LUFS (target ${I}, Δ${parseFloat(deltaI) >= 0 ? '+' : ''}${deltaI}) | TP=${stats.input_tp} dBTP (ceiling ${TP}) | LRA=${stats.input_lra} (target ${LRA}) | thresh=${stats.input_thresh} | measure=${pass1Duration}s`);
 
+    if (!needsLoudnormCorrection(stats)) {
+        console.log(`[loudnorm] ${filename}: within ${LOUDNORM_TOLERANCE_LU} LU of the ${I} LUFS target, skipping normalization`);
+        return;
+    }
```
Comment (kouloumos) [inline src/tasks/downloadYTV.ts:194]: Informative, no change needed. This early return skips pass 2, which was also the only step re-encoding the audio to AAC 128k -- so the two branches now differ: a file that needs correction gets re-encoded, one within tolerance keeps its original audio untouched. For YouTube sources that's already AAC so it's a no-op, but a CDN/direct-URL source that happens to land on-target now passes its original codec/bitrate straight through. Harmless in practice -- just flagging the asymmetry so it's a conscious tradeoff and not a surprise later.

**Review by kouloumos — APPROVED** [review]
Looks good. The two-pass measurement regression is real and `-vn` on the input is the right fix; the skip gate is well-reasoned and I like that the TP-vs-loudness tradeoff is documented in the code, the tests, and the field report on the PR. Both bot comments are addressed -- verified the `Number.isFinite` guard and the input-side `-vn` both landed.

Left one inline note on the skip path -- informative, no action needed, just a behavioral asymmetry worth being aware of. Details there.
